### PR TITLE
Modernize dependencies

### PR DESCRIPTION
**Pull request by rtholmes**
# Status
* Dependencies updated
* CircleCI updated

# Overview
PR for fixing #428 

- [x] Node - existing: v12; current: v18 is LTS until 2025-05.
- [x] Typescript - existing: 3.7.2; current: v4.8.4.
- [x] Mongo - existing: 3.0.2; current: v5 is LTS until 2025. NOTE: just using the `mongo:latest` docker image.
- [x] Restify v10

## Notes worth knowing

* `node-fetch` has to stay at v2.X; v3 only supports ESM modules (which we don't want to transition to).

### CODE REVIEW

**Comment by rtholmes** [comment]
Currently broken because of restify node18 problem. 

Issue: https://github.com/restify/node-restify/issues/1925
PR: https://github.com/restify/node-restify/pull/1906

**Comment by winstan** [comment]
Restify supports node18 today via:
PR: https://github.com/restify/node-restify/pull/1929

**Comment by rtholmes** [comment]
@winstan This seems to be ready to try on a host. I can't merge to master because 310 is still taking changes and I don't want to risk the changes to the DB/Node this late in the term.

Is there a dev instance somewhere we can spin up from this PR?

File: packages/proxy/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM nginx:1.13-alpine
+FROM node:lts-alpine
```
Comment (winstan) [inline packages/proxy/Dockerfile:1]: I don't think you meant to change the container from nginx to node... I think you meant to change it to this?
FROM nginx:stable-alpine

Comment (rtholmes) [reply packages/proxy/Dockerfile]: fixed, good catch!

**Comment by rtholmes** [comment]
Also @winstan, if you happen to know what mongo image to use other than `mongo:latest`, I really struggled for the right version. v5 is in LTS, but v6 is out: it would be nice to have a v5latest image, but I couldn't find one.

**Comment by winstan** [comment]
> Also @winstan, if you happen to know what mongo image to use other than `mongo:latest`, I really struggled for the right version. v5 is in LTS, but v6 is out: it would be nice to have a v5latest image, but I couldn't find one.

Sorry, nope. :(

**Comment by winstan** [comment]
Dev hosts: classy-dev and cs210-dev. I've given you access to both. (Talk to Paul about cs210-dev.)
Be sure to use /opt/classy-scripts/fix-classy to reset permissions on /var/opt/classy so you can make changes (like nuke the database)
I got it to build on classy-dev, but the database isn't happy...

**Comment by winstan** [comment]
If you want a cs310-dev host, send a request to help@cs...

**Comment by rtholmes** [comment]
> > Also @winstan, if you happen to know what mongo image to use other than `mongo:latest`, I really struggled for the right version. v5 is in LTS, but v6 is out: it would be nice to have a v5latest image, but I couldn't find one.
> 
> Sorry, nope. :(

I almost wonder if we want to use `mongo:5.0.14`, just so we can be in control of updating the mongo image? I'm fine with `node:lts-alpine` since it's v18 and should be for a few years.

**Comment by winstan** [comment]
Playing some more with getting the db to run... Complained of missing processor extension AVX. I found a way to update the processor model for the VM and now everything seems to start up and run fine. (classy-dev)
I need to update the processor model for all Classy boxes now...

**Comment by rtholmes** [comment]
@winstan glad you got it going! I'm running a modern mongo on my local dev instance so the db behaviours should be ok, but I never thought about that kind of problem! I do know they changed the auth mechanism though (there have been warnings on the terminal for a while) so I'll have to look into those (next week) and see whether I need to change how we open connections to mongo.

**Comment by rtholmes** [comment]
@winstan I moved 310 onto a branch so this is probably good to merge to master. This might also be a good time (after the merge) to rename `master` to `main`, is that ok with you?

**Comment by winstan** [comment]
> @winstan I moved 310 onto a branch so this is probably good to merge to master. This might also be a good time (after the merge) to rename `master` to `main`, is that ok with you?

Sure, sounds good.

**Pull request merged**